R/gradethis_equal.R

```diff
@@ -0,0 +1,25 @@
+#' Compare two `ggplot`s to check whether they are equal
+#'
+#' @param x,y Two `ggplot` objects to compare
+#' @param ... Unused
+#'
+#' @seealso [gradethis::gradethis_equal()] for the generic function.
+#' @inherit gradethis::gradethis_equal return
+#' @importFrom gradethis gradethis_equal
+#' @export
+#'
+#' @examples
+#' library(ggplot2)
+#' library(ggcheck)
+#' library(gradethis)
+#'
+#' cty_plot <- ggplot(mpg, aes(x = displ, y = cty)) + geom_point()
+#' hwy_plot <- ggplot(mpg, aes(x = displ, y = cty)) + geom_point()
+#'
+#' gradethis_equal(cty_plot, hwy_plot)
+#' gradethis_equal(cty_plot, cty_plot)
+gradethis_equal.ggplot <- function(x, y, ...) {
+  try(ggplot2::ggplot_build(x), silent = TRUE)
+  try(ggplot2::ggplot_build(y), silent = TRUE)
```

Comment:
Is `ggplot_build()` a fairly quick operation? Regardless, this looks necessary for comparison.

Reply:
It's a whole lot faster than my first idea, which was to `print()` both plots and capture the output so it isn't shown to the student:

``` r
library(ggplot2)

plot1 <- ggplot(mpg, aes(displ, hwy, colour = class)) + 
  geom_point()
plot2 <- ggplot(mpg, aes(displ, hwy, colour = class)) + 
  geom_point()

bench::mark(
  ggplot_build(plot1),
  print(plot2),
  check = FALSE
)

    #> # A tibble: 2 × 6
    #>   expression               min   median `itr/sec` mem_alloc `gc/sec`
    #>   <bch:expr>          <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
    #> 1 ggplot_build(plot1)   21.6ms   25.3ms     39.2     4.67MB     13.7
    #> 2 print(plot2)         412.1ms  417.7ms      2.39    4.61MB     10.8
```

<sup>Created on 2023-05-05 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

But yeah, I think at least doing this step is unavoidable. The alternative would be manually ignoring the computed fields inside the `ggplot` object, but that feels much less robust, since there's always a chance I'll miss some obscure element that only gets computed for one specific geom

Reply:
Yeah, this makes sense. `matplotlib` in Python works in a similar way where you can prep a plot to be displayed. Interesting to see a similar thing here for `ggplot`.